[CORE-13383] Implement KIP-709: offset fetch v8

BenPope opened the pull request:
This PR implements KIP-709 for offset fetch version 8, enabling support for fetching offsets for multiple consumer groups in a single request. Currently, the implementation runs tests against version 7 to verify existing functionality isn't broken.

Key changes:
- Extends the offset fetch protocol to support multiple groups in a single request
- Updates data structures to use chunked_vector for better memory efficiency
- Adds new protocol schema definitions for v8 request/response structures


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Features

* Support KIP-709

BenPope committed:
kafka/protocol: Update OffsetFetch schema to v8

Pin cluster_link to v7 for now, since that's what it supports.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/protocol: Update OffsetFetch to use chunked_vector

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/protocol: Refactor Make OffsetFetchResponse constructors

No functional change.

* Make the constructors explicit
* Take the sink parameter by value, to make the callsite obvious
* Extract get_topics to a general function

Signed-off-by: Ben Pope <ben@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73193-->
<!--af3f7859-c6b1-4678-8017-a95d080f67f8-->
<!--fb169811-6fec-4ac0-8e18-44e0e93ca999-->
<!--ecbe8c60-89e0-4c82-9ad2-6a83a3b18e7d-->
<!--0f6c7a4e-61a1-4286-8114-3442f71c9936-->
<!--d6375897-1207-407f-ad24-0e8eca8e06c4-->
<!--26d8e359-9caf-4993-92e8-adfce2ed8b9d-->
<!--fe31d31b-bdd3-45c8-b794-d2006db3c3db-->
<!--title-->
## Retry command for [Build#73193](https://buildkite.com/redpanda/redpanda/builds/73193)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":false,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":true,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_consumer_groups_mirroring@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_continuous_group_sync@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"},"with_failures":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--41b5551c-973a-4d9f-9be4-96e76bdea86d-->
<!--TEST RESULTS-->
<!--b2e4863e-a243-463e-979d-6d2795e262bc-->
<!--TEST RESULTS-->
<!--7b71563f-8ada-4726-88f7-3b107d6daaa3-->
<!--TEST RESULTS-->
<!--cbfe97b5-d7a6-4de4-a6cd-803add622a71-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73193">build#73193</a></summary>


| test_class                           | test_method                        | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                                         |
|:-------------------------------------|:-----------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring     | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                              | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f37-4c8c-9b18-86a93bba7342 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring    |
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring     | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                              | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f6-454e-bcc1-2cb0fbc58d08 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring    |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync         | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f39-4a01-a5e2-667390119746 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync        |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync         | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f8-4e2b-bc6d-b6529bd656d0 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync        |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f34-45b8-906d-3ac0f4adf2fb | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f4-457c-bd3f-c364147673cd | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f37-4c8c-9b18-86a93bba7342 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f6-454e-bcc1-2cb0fbc58d08 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ConsumerGroupTest                    | test_group_recovery                | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f35-4209-92fc-4298c36792f7 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_group_recovery                                  |
| ConsumerGroupTest                    | test_group_recovery                | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f4-4b28-b627-6b3ba62ccb6b | FAIL          | 0/21     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_group_recovery                                  |
| WriteCachingFailureInjectionE2ETest  | test_crash_all_with_consumer_group | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978b-6f34-45b8-906d-3ac0f4adf2fb | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all_with_consumer_group |
| WriteCachingFailureInjectionE2ETest  | test_crash_all_with_consumer_group | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73193#0199978f-18f4-457c-bd3f-c364147673cd | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all_with_consumer_group |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73226">build#73226</a></summary>


| test_class                           | test_method                        | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                                         |
|:-------------------------------------|:-----------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring     | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                              | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring    |
| ShadowLinkConsumeGroupsMirroringTest | test_consumer_groups_mirroring     | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                              | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_consumer_groups_mirroring    |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync         | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync        |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync         | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync        |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ShadowLinkingReplicationTests        | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                   |
| ConsumerGroupTest                    | test_group_recovery                | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_group_recovery                                  |
| ConsumerGroupTest                    | test_group_recovery                | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupTest&test_method=test_group_recovery                                  |
| WriteCachingFailureInjectionE2ETest  | test_crash_all_with_consumer_group | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a86-898a-4fe5-835c-6e36a40b87e8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all_with_consumer_group |
| WriteCachingFailureInjectionE2ETest  | test_crash_all_with_consumer_group | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73226#01999a87-1dca-4fd3-b799-821598327d38 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all_with_consumer_group |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73332">build#73332</a></summary>


| test_class                    | test_method                        | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:------------------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic             | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73332#01999fd4-164d-4e90-8624-acc0e627cc4f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic     |
| DataMigrationsApiTest         | test_migrated_topic_data_integrity | {"include_groups": false, "params": {"cancellation": {"dir": "in", "stage": "prepared"}, "use_alias": false}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/73332#01999fcf-4fcf-4737-ae15-ab7a9b2f4c9c | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_migrated_topic_data_integrity |
| RapidTopicRecreateTest        | test_topic_rapid_recreation        | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73332#01999fcf-4fd5-4981-b328-3107e95743a9 | FLAKY         | 20/21    | upstream reliability is '99.50372208436724'. current run reliability is '95.23809523809523'. drift is 4.26563 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73626">build#73626</a></summary>


| test_class                    | test_method                                  | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------------|:---------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic                       | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73626#0199b9b7-eb1f-4ad8-b575-becb4227ffd1 | FLAKY         | 19/21    | upstream reliability is '96.21212121212122'. current run reliability is '90.47619047619048'. drift is 5.73593 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| SegmentMsTest                 | test_segment_rolling_with_retention_consumer | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73626#0199b9b2-7e1f-475b-afdd-cd5a36042862 | FLAKY         | 16/21    | upstream reliability is '98.76237623762376'. current run reliability is '76.19047619047619'. drift is 22.5719 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<!--/body-->

BenPope commented:
/ci-repeat 1
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":false,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":true,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_consumer_groups_mirroring@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_continuous_group_sync@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"},"with_failures":false}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73226-->
<!--74eab348-4a00-48db-aa59-501f56f4293e-->
<!--ec542d18-4bd2-4d76-8bd5-e006466fdb25-->
<!--title-->
## Retry command for [Build#73226](https://buildkite.com/redpanda/redpanda/builds/73226)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_consumer_groups_mirroring@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":false,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/consumer_group_test.py::ConsumerGroupTest.test_group_recovery
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":true,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkConsumeGroupsMirroringTest.test_continuous_group_sync@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"},"with_failures":false}
```
<!--/body-->

BenPope committed:
kafka/server: Support offset_fetch v8

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/server: Refactor offset_fetch

No functional change.

* Extract functionality into functions.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/server: Reimplement offset_fetch

In preperation for offset_fetch v8, which allows multiple groups
to be requested, implement everything as if it were v8.

If the incoming request is < v8, then convert the request to latest,
and the response to the version that was requested.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
kafka/server: Refactor offset_fetch

No functional change.

* Reorder functions

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented on src/v/kafka/client/consumer.cc:
```diff
@@ -369,11 +369,11 @@ ss::future<describe_groups_response> consumer::describe_group() {
     auto req_builder = [topics{std::move(topics)}, group_id{_group_id}]() {
         return offset_fetch_request{
-          .data{.group_id = group_id, .topics = topics}};
+          .data{.group_id = group_id, .topics = topics.copy()}};
```
They're needed for retries, no change in behaviour here, the copy was already there.

BenPope commented on src/v/kafka/client/client.cc:
```diff
@@ -604,11 +604,11 @@ client::consumer_assignment(const group_id& g_id, const member_id& name) {
     return get_consumer(g_id, name)
       .then([this, topics{std::move(topics)}](shared_consumer_t c) mutable {
           return gated_retry_with_mitigation([c, topics{std::move(topics)}]() {
-              return c->offset_fetch(topics);
+              return c->offset_fetch(topics.copy());
```
Similar response, it's needed and was already there.

BenPope committed:
kafka/server: Add offset_fetch testing

Signed-off-by: Ben Pope <ben@redpanda.com>

joe-redpanda commented on src/v/kafka/protocol/offset_fetch.h:
```diff
@@ -50,35 +50,46 @@ struct offset_fetch_response final {
-        data.error_code = error_code::none;
-        if (topics) {
+    template<typename Topics>
+    static auto get_topics(auto topics) {
```
nice

joe-redpanda commented on src/v/kafka/server/tests/consumer_groups_test.cc:
```diff
@@ -223,17 +223,28 @@ FIXTURE_TEST(offset_commit_and_fetch_request, consumer_offsets_fixture) {
+            if (api_version < version_with_groups) {
+                check_response_topics(resp.data, all_topics);
+            } else {
+                BOOST_REQUIRE_EQUAL(resp.data.groups.size(), 1);
```
Can you add a test case with >1 groups?

BenPope replied:
Dine

BenPope replied:
> Can you add a test case with >1 groups?

Done

joe-redpanda commented on src/v/kafka/server/server.cc:
```diff
@@ -1085,92 +1086,202 @@ ss::future<response_ptr> add_partitions_to_txn_handler::handle(
     });
 }
 
+namespace {
```
Nit: can we take this opportunity to split this into a separate handler file?

BenPope replied:
KIP-848 is coming soon; I think that's the best opportunity for reorganisation.

joe-redpanda reviewed (dismissed):
Minor comments, otherwise looks good

mmaslankaprv commented on src/v/kafka/server/group_router.cc:
```diff
@@ -260,8 +261,46 @@ group_router::leave_group(leave_group_request&& request) {
+group_router::offset_fetch(offset_fetch_request request) {
+    chunked_hash_map<
+      std::optional<std::pair<model::ntp, ss::shard_id>>,
+      offset_fetch_request>
+      requests_by_shard;
+
+    offset_fetch_response response;
+    // Collect requests by shard
+    for (auto& group : request.data.groups) {
+        requests_by_shard
+          .try_emplace(shard_for(group.group_id), offset_fetch_request{})
+          .first->second.data.groups.push_back(std::move(group));
+    }
+    // Collect responses
+    for (auto& [key, request] : requests_by_shard) {
+        if (key.has_value()) {
+            auto& [ntp, shard] = key.value();
+            request.ntp = std::move(ntp);
+            auto shard_res = co_await with_scheduling_group(
+              _sg, [this, shard, r = std::move(request)]() mutable {
+                  return get_group_manager().invoke_on(
+                    shard, _ssg, &group_manager::offset_fetch, std::move(r));
+              });
+            std::ranges::move(
+              shard_res.data.groups, std::back_inserter(response.data.groups));
+        } else {
+            // Not coordinator for these groups
+            for (auto& group : request.data.groups) {
+                vlog(
+                  cg_klog.trace,
+                  "in route() not coordinator for {}",
+                  group.group_id);
+                response.data.groups.push_back(
+                  offset_fetch_response_group{
+                    .group_id = std::move(group.group_id),
+                    .error_code = error_code::not_coordinator});
+            }
+        }
+    }
+    co_return response;
```
this is neat

BenPope replied:
> this is neat

I think there's some kind of map_reduce abstraction in here.

mmaslankaprv commented on src/v/kafka/server/server.cc:
```diff
@@ -1102,6 +1102,61 @@ auto partition_auth_describe(
+    vassert(
+      res.data.groups.size() == 1,
+      "Only single group fetch is supported in v{}",
+      target_version);
```
I would remove this assertion as it depends on user input ?

BenPope replied:
> I would remove this assertion as it depends on user input ?

It depends on:
* Deserialization
* The logic for offset fetch

It shouldn't be possible to have not 1 group in version <= 8, but I've changed it from an assert anyway.

BenPope committed:
kafka/server: Add offset_fetch AuthZ testing

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ceb254cb6d35a06176abda033e2f045213b76af8..4555fe53131ebbe2ce49ed0ccc24a764ee58e5e6)
* Do not assert if there is not one group
* Add tests for AuthZ and multiple groups (tests are rewritten)
* Fix incorrect error code for topic auth.
